### PR TITLE
[query] Add capability for parallel block iteration

### PR DESCRIPTION
**Pull request by arnikola**
**What this PR does / why we need it**:

Should reduce user impact seen in #1678 

**Special notes for your reviewer**: Majority of work is in `encoded_step_iterator_generic.go`, `encoded_unconsolidated_step_iterator`, and `encoded_step_iterator`; rest of is plumbingg

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:
```documentation-note
NONE
```

### CODE REVIEW

File: src/query/ts/m3db/consolidators/step_accumulator.go [hunk 1]
```diff
@@ -61,35 +56,28 @@ func NewStepLookbackAccumulator(
...
 	// iteration of the datapoint set.
 	c.earliestLookback = c.earliestLookback.Add(c.stepSize)
-	accumulated := make([]xts.Datapoints, len(c.datapoints))
-	for i, dps := range c.datapoints {
-		accumulated[i] = make(xts.Datapoints, len(dps))
-		copy(accumulated[i], dps)
-		c.datapoints[i] = c.datapoints[i][:0]
-	}
-
+	accumulated := make([]xts.Datapoint, len(c.datapoints))
```
Comment (robskillington) [inline src/query/ts/m3db/consolidators/step_accumulator.go]: Is it plausible to pool these allocations at some point in the future?  

I'm still trying to understand the lifecycle better here, so I'm definitely out of my depth in being able to say one way or another whether it's a good idea hah.

File: src/query/ts/m3db/consolidators/step_accumulator.go [hunk 2]
```diff
@@ -35,7 +35,7 @@ type StepLookbackAccumulator struct {
 	lookbackDuration time.Duration
 	stepSize         time.Duration
 	earliestLookback time.Time
-	datapoints       [][]xts.Datapoint
+	datapoints       []xts.Datapoint
```
Comment (robskillington) [inline src/query/ts/m3db/consolidators/step_accumulator.go]: Nice, this was able to be flattened?

File: src/query/ts/m3db/encoded_step_iterator.go [hunk 3]
```diff
@@ -26,43 +26,70 @@ import (
...
-		cs.currentTime,
-		len(iters),
-		cs.consolidationFn,
+	var (
+		cs       = b.consolidation
+		iters    = b.seriesBlockIterators
+		lookback = b.options.LookbackDuration()
+
+		collectors       = make([]*consolidators.StepLookbackConsolidator, len(iters))
+		seriesCollectors = make([]consolidators.StepCollector, 0, len(iters))
```
Comment (robskillington) [inline src/query/ts/m3db/encoded_step_iterator.go]: nit: Maybe make them both size zero but cap len(iters) or the opposite so it's consistent?

File: src/query/ts/m3db/encoded_step_iterator_generic.go [hunk 4]
```diff
@@ -88,43 +101,103 @@ func (it *encodedStepIterWithCollector) nextForStep(
...
+	for i := range it.seriesIters {
+		var (
+			i         = i
+			peek      = it.seriesPeek[i]
+			iter      = it.seriesIters[i]
+			collector = it.seriesCollectors[i]
+		)
+
+		it.workerPool.Go(func() {
+			peek, collector, err := nextForStep(peek, iter, collector, stepTime)
```
Comment (robskillington) [inline src/query/ts/m3db/encoded_step_iterator_generic.go]: Nice this is a solid approach.

I wonder if it would also be a good idea to basically do a bunch in parallel (like the next X points defined by a constant, like 10 or 20) and then read from intermediary storage after?  That way we don't have to use tons of memory decoding everything up front but every time we do a decode step, we can hot loop the decoding for a little while.  Would also help with less context switching (using goroutines to do decoding, then waiting on result less often).

Basically exactly what you're doing here in batch.

File: src/query/ts/m3db/encoded_step_iterator_generic.go [hunk 5]
```diff
@@ -88,43 +101,103 @@ func (it *encodedStepIterWithCollector) nextForStep(
...
 	if it.err != nil {
 		return false
 	}
 
-	it.stepTime = it.stepTime.Add(bounds.StepSize)
-	return !bounds.End().Before(it.stepTime)
+	it.stepTime = it.stepTime.Add(it.meta.Bounds.StepSize)
+	next := !it.blockEnd.Before(it.stepTime)
+	if next && it.updateFn != nil {
+		it.updateFn()
```
Comment (robskillington) [inline src/query/ts/m3db/encoded_step_iterator_generic.go]: What's this used for btw?

File: src/query/ts/m3db/encoded_step_iterator_test.go [hunk 6]
```diff
@@ -268,3 +310,159 @@ func TestConsolidatedStepIteratorSplitByBlock(t *testing.T) {
...
+		}
+
+		for _, it := range pIters {
+			for _, p := range points {
+				it.EXPECT().Current().Return(p, xtime.Second, annotation)
+			}
+
+			it.EXPECT().Next().DoAndReturn(func() bool {
+				// simulate load
+				time.Sleep(time.Millisecond)
```
Comment (robskillington) [inline src/query/ts/m3db/encoded_step_iterator_test.go]: Hm instead of mock iterators how about just encoding a bunch into real M3TSZ encoders?  

Not sure if time sleeping is a good proxy for real decoding work since always waiting for a fixed amount of time at each point will have different scheduling characteristics vs actually doing CPU work during a call to Next() on a single iterator.

File: src/query/ts/m3db/consolidators/step_accumulator.go (hunk 2 above)
Comment (arnikola) [reply src/query/ts/m3db/consolidators/step_accumulator.go]: Yeah, I was initially trying to abstract away the `i-th` wise stuff in the generic step iterator and let this accumulator handle it, but that doesn't lend very well to parallelization

File: src/query/ts/m3db/encoded_step_iterator.go (hunk 3 above)
Comment (arnikola) [reply src/query/ts/m3db/encoded_step_iterator.go]: Sure, is there any preference either way? I've generally leaned toward initializing explicitly with size when I know what the size will be, but not sure that's best (seems `append` is often a little safer)

File: src/query/ts/m3db/encoded_step_iterator_generic.go (hunk 5 above)
Comment (arnikola) [reply src/query/ts/m3db/encoded_step_iterator_generic.go]: This collates the results post `Next()`, so when `Current()` is called on the iterator it does virtually no work. Also partially necessary here because the step collector `ConsolidateAndMoveToNext` functions currently mutate their internal steps, which is easy enough to change but this way tries to avoid the extra checks etc that would be added otherwise

File: src/query/ts/m3db/encoded_step_iterator_test.go (hunk 6 above)
Comment (arnikola) [reply src/query/ts/m3db/encoded_step_iterator_test.go]: Yeah totally fair, i'll try and do one with no mocks

File: src/query/ts/m3db/consolidators/step_accumulator.go (hunk 1 above)
Comment (arnikola) [reply src/query/ts/m3db/consolidators/step_accumulator.go]: Slightly simplified flow for this is that the `N-th` iterator in the encoded step iterator has it's `Next()` called, adding those points to the `N-th StepCollector`, until it either

a) returns false, in which case the accumulated values are consolidated by the `StepCollector` and set to the `Current` values

b) yields a point whose timestamp is beyond the current step, in which case that point is saved as a "peeked value" in preparation for the next time `Next()` is called, and the accumulated values are consolidated as for the above case.

We definitely should add pooling here but it may be a fairly involved change largely because we don't have a great lifecycle management story for them yet

Comment (arnikola) [reply src/query/ts/m3db/consolidators/step_accumulator.go]: In terms of the different `StepCollectors`, the step iterator variant that uses `StepLookbackAccumulator` is the 
"unconsolidated" one, which just gives back all the raw points. These are used only in temporal functions like `sum_over_time`, since they need to work on the rawest data available. All othe functions use the `StepLookbackConolidator`, which gives "consolidated" data points, which are snapped to stepSize and have a given consolidation function applied to them.

Frankly, I really dislike the concept of having this consolidated/unconsolidated split as is; I think we may be able to add a tiny bit of  carefully placed logic to slot in a specialized consolidation function where necessary (also that will  also give us `summarize` for free).

File: src/query/ts/m3db/encoded_step_iterator.go (hunk 3 above)
Comment (robskillington) [reply src/query/ts/m3db/encoded_step_iterator.go]: Append just tends to be safer yeah, I've seen a lot of bugs of creating size, then using append and then empty values sit in front of real values in the result.

If it's super straight forward and in super hot path it's probably justified to not use append (since might be slightly faster, don't need to keep updating the size of the array in the slice, although that's obviously super fast addition too).
